find_each rather than find (#182)
andycroll/andycroll.com: _posts/2023/2023-02-27-consider-find_each-for-active-record-assocation-loops.md

---
title: "Consider find_each for looping through Active Record associations"
description: "Rarely used, but protects you against foot shooting"
layout: article
category: ruby
image:
  base: '2023/consider-find_each-for-looping-through-active-record-associations'
  alt: "Light loops"
  credit: "Nareeta Martin"
  source: "https://unsplash.com/photos/za2gMSfLmzU"

---

> The standard method for enumerating through groups of objects is `each`, both through arrays in Ruby and through Active Record models in Rails.

Reviewer: I would bout "both through arrays....Rails" after "group of objects", therefore ending the sentence with: "is `each`".

Author: ```suggestion
The standard method for enumerating through groups of objects, both through arrays in Ruby and through Active Record models in Rails, is `each`.
```

> However, if you are looping over all the records for a model (perhaps to backfill data) or even over a large association you can encounter severe memory and speed issues both loading and processing enormous amounts of records as you loop.

Reviewer: *if you are looping over a large amount of data, perhaps looping over all of the records for a model in order to backfill data, you can encounter...., both with loading and processing each record.

Author: ```suggestion
However, if you are looping over a large amount of data, perhaps all the records for a model in order to backfill data, you can encounter severe memory and speed issues, both with loading and processing large volumes of records.
```

> You should consider the functionality provided by [Active Record’s Batches](https://api.rubyonrails.org/classes/ActiveRecord/Batches.html), we’ll demonstrate using `find_each`.

Reviewer: ```suggestion
You should consider the functionality provided by [Active Record’s Batches](https://api.rubyonrails.org/classes/ActiveRecord/Batches.html). We’ll demonstrate using `find_each`.
```

## Instead of…

…looping through lots of Active Record objects using `.each`:

```ruby
post.comments.each do |comment|
  # Do stuff with each comment: enqueue a job
end
```

## Use…

> 

Reviewer: Don't repeat 'use...'

> …use `.find_each` to more efficiently load records from the database:

Author: ```suggestion
…`.find_each` to more efficiently load records from the database:
```

```ruby

> post.comments.find_each do |comment|

Reviewer: I did this today when needing to re-loop through all TSG user models to reprocess image variants! :-D

>   # Do stuff with each comment: enqueue a job?

Reviewer: Why a question mark after "enqueue a job" here but not earlier?

I'd also just change it to "e.g. enqueue a job".

Author: ```suggestion
  # Do stuff with each comment: enqueue a job
```

end
```

## Why?

> Using `each` makes _one_ SQL call to the database and tries to load the entire set of objects into memory and then loop over them. It’s the same as calling `post.comments.all.each` in the first example.

Reviewer: ```suggestion
Using `each` makes _one_ SQL call to the database and tries to load the entire set of objects into memory and then loop over them. It’s the same as if you'd called `post.comments.all.each` instead in the first example.
```

> This is problem in two dimensions. Firstly the database query may take a long time to execute or may timeout. Secondly, when (and if) it does, there’s likely to be significant memory usage as it loads all the records into memory in order to loop over them. 

Reviewer: *is a problem

*First...

*time out

* Second..

*when (_or_ if), there's likely to be significant memory usage as your programme attempts to load...

Author: ```suggestion
This is a problem in two dimensions. First the database query may take a long time to execute or may time out. Second, when (or if) it does return data, there’s likely to be significant memory usage as it loads all the records into memory in order to loop over them. 
```

> This is a problem in two dimensions. First the database query may take a long time to execute or may time out. Second, when (or if) it does return data, there’s likely to be significant memory usage as it loads all the records into memory in order to loop over them. 

Reviewer: *First,

Author: ```suggestion
This is a problem in two dimensions. First, the database query may take a long time to execute or may time out. Second, when (or if) it does return data, there’s likely to be significant memory usage as it loads all the records into memory in order to loop over them. 
```

> Using `.find_each` makes a series of more efficient SQL queries (with a bunch of sensible defaults) and uses a cursor-based approach to retrieve records from the database, which can be more efficient than loading all records into memory at once.

Reviewer: Don't know what 'cursor-based approach' means. Is it commonly-understood?

*which is _often_ (?) much/a lot more efficient than...

Author: ```suggestion
Using `.find_each` makes a series of more efficient SQL queries (with a bunch of sensible defaults) to retrieve records from the database, which is often a lot more efficient than loading all records into memory at once.
```

## Why not?

> At a certain size you’ll need to move on from looping over every record.

Reviewer: This isn't a "Why not?", right?

This is perhaps for the section above? Almost like: well, at a certain point you'll _have_ to do this!

Author: No, this is because even _this_ won't work once you're at a certain scale!

Author: ```suggestion
At large scale you’ll need to use even more advanced techniques and you’ll have to move beyond long running loops over every record. Perhaps to reduce a high volume of read/write operations on your database, perhaps to try and get this loop to run in minutes rather than hours!
```

> At large scale you’ll need to use even more advanced techniques and you’ll have to move beyond long running loops over every record. Perhaps to reduce a high volume of read/write operations on your database, perhaps to try and get this loop to run in minutes rather than hours!

Reviewer: ```suggestion
At large scale you’ll need to use even more advanced techniques and you’ll have to move beyond long running loops. Perhaps you want to reduce a high volume of read/write operations on your database or you want to cut down execution time from minutes to hours!
```

I this this point should be the last "Why not?" point.

If you need to see the records in a particular order, `.find_each` doesn't support that.

> If you need to modify the records in place this sort of looping isn’t ideal. For example, if you’re running an `#update` on each record, you could consider other more appropriate methods like `#update_all`.

Reviewer: Why isn't it ideal?

Author: ```suggestion
If you need to modify the records in place this sort of looping isn’t ideal. For example, if you’re running an `#update` on each record, you’ll be executing _a lot_ of queries, you could consider other more appropriate methods for bulk updating methods such as `#update_all`.
```

> If you need to modify the records in place this sort of looping isn’t ideal. For example, if you’re running an `#update` on each record, you’ll be executing _a lot_ of queries, you could consider other more appropriate methods for bulk updating methods such as `#update_all`.

Reviewer: *place,

Reviewer: * _a lot_ of queries. Instead consider more appropriate methods for... , such as `#update_all`.

> You should also be aware that if you’re looping over all the associated records you should _not_ be doing that to send it to view. Having an unbounded loop in view code when you don't know how many records you’ll have is a recipe for enormous pages and poor user experience, even if you do use `.find_each`. Consider [using pagination](/ruby/do-not-use-all-without-pagination-or-limit).

Reviewer: Maybe start this paragraph with: Don't use this to generate view code, either the `each` or `find_each` variation. Having an unbounded loop....is a recipe for slow loading pages and a poor user experience. Consider using pagination instead.

Author: ```suggestion
Don't use this to generate view code, either the `each` _or_ `find_each` version. Having an unlimited loop in view code when you don't know how many records you’ll have is a recipe for slow pages and poor user experience. Consider [using pagination](/ruby/do-not-use-all-without-pagination-or-limit) instead.
```

> Don't use this to generate view code, either the `each` _or_ `find_each` version. Having an unlimited loop in view code when you don't know how many records you’ll have is a recipe for slow pages and poor user experience. Consider [using pagination](/ruby/do-not-use-all-without-pagination-or-limit) instead.

Reviewer: *Don't use either the `each` or `find_each` version to generate...

*Having a potentially incredibly large (not sure if that's right but I think we know it's not 'unlimited' in most cases) loop, given you don't know how many records you'll have, is a...

> On the other end of the scale the Rails guides suggest `find_each` is only for processing a large number of records that wouldn't fit in memory all at once. If you just need to loop over less than “a thousand” records the regular methods are recommended.

Reviewer: ```suggestion
The Rails guides suggest `find_each` is only needed for processing a large number of records that wouldn't fit in memory all at once. If you only need to loop over fewer than one thousand records the regular methods are fine/recommended.
```

> The Rails guides suggest `find_each` is only needed for processing a large number of records that wouldn't fit in memory all at once. If you only need to loop over fewer than one thousand records the regular methods are fine/recommended.

Reviewer: Hehe, I was giving you an option between 'fine' or 'recommended', but I think leaving it like this does work!
